File: java/src/main/java/benchmarks/Compression.java
package benchmarks;

import java.util.*;
import java.nio.ByteBuffer;

public class Compression extends Benchmark {
    private int iterations;
    private byte[] testData;
    private long result;
    
    // ==================== BWT ====================
    private static class BWTResult {
        byte[] transformed;
        int originalIdx;
        
        BWTResult(byte[] transformed, int originalIdx) {
            this.transformed = transformed;
            this.originalIdx = originalIdx;
        }
    }
    
    private BWTResult bwtTransform(byte[] input) {
        int n = input.length;
        if (n == 0) {
            return new BWTResult(new byte[0], 0);
        }

        // 1. Создаём суффиксный массив
        Integer[] sa = new Integer[n];
        for (int i = 0; i < n; i++) {
            sa[i] = i;
        }

        // 2. Фаза 0: сортировка по первому символу (Radix sort)
        List<Integer>[] buckets = new List[256];
        for (int i = 0; i < 256; i++) {
            buckets[i] = new ArrayList<>();
        }
        
        for (int idx : sa) {
            int firstChar = input[idx] & 0xFF;
            buckets[firstChar].add(idx);
        }

        int pos = 0;
        for (int b = 0; b < 256; b++) {
            for (int idx : buckets[b]) {
                sa[pos++] = idx;
            }
        }

        // 3. Фаза 1: сортировка по парам символов
        if (n > 1) {
            // Присваиваем ранги по первому символу
            int[] rank = new int[n];
            int currentRank = 0;
            int prevChar = input[sa[0]] & 0xFF;
            
            for (int i = 0; i < n; i++) {
                int idx = sa[i];
                int currChar = input[idx] & 0xFF;
                if (currChar != prevChar) {
                    currentRank++;
                    prevChar = currChar;
                }
                rank[idx] = currentRank;
            }

            // Сортируем по парам (ранг[i], ранг[i+1])
            int k = 1;
            while (k < n) {
                // Создаём пары
                int[][] pairs = new int[n][2];
                for (int i = 0; i < n; i++) {
                    pairs[i][0] = rank[i];
                    pairs[i][1] = rank[(i + k) % n];
                }

                // Сортируем индексы по парам
                Arrays.sort(sa, new Comparator<Integer>() {
                    @Override
                    public int compare(Integer a, Integer b) {
                        int[] pairA = pairs[a];
                        int[] pairB = pairs[b];
                        if (pairA[0] != pairB[0]) {
                            return Integer.compare(pairA[0], pairB[0]);
                        }
                        return Integer.compare(pairA[1], pairB[1]);
                    }
                });

                // Обновляем ранги
                int[] newRank = new int[n];
                newRank[sa[0]] = 0;
                for (int i = 1; i < n; i++) {
                    int[] prevPair = pairs[sa[i - 1]];
                    int[] currPair = pairs[sa[i]];
                    if (prevPair[0] != currPair[0] || prevPair[1] != currPair[1]) {
                        newRank[sa[i]] = newRank[sa[i - 1]] + 1;
                    } else {
                        newRank[sa[i]] = newRank[sa[i - 1]];
                    }
                }

                rank = newRank;
                k *= 2;
            }
        }

        // 4. Собираем BWT результат
        byte[] transformed = new byte[n];
        int originalIdx = 0;

        for (int i = 0; i < n; i++) {
            int suffix = sa[i];
            if (suffix == 0) {
                transformed[i] = input[n - 1];
                originalIdx = i;
            } else {
                transformed[i] = input[suffix - 1];
            }
        }

        return new BWTResult(transformed, originalIdx);
    }    
    // Быстрое обратное BWT как в Rust
    private byte[] bwtInverse(BWTResult bwtResult) {
        byte[] bwt = bwtResult.transformed;
        int n = bwt.length;
        
        if (n == 0) {
            return new byte[0];
        }
        
        // Оптимизация: используем примитивные массивы как в Rust
        // 1. Подсчитываем частоты символов
        int[] counts = new int[256];
        for (byte b : bwt) {
            counts[b & 0xFF]++;
        }
        
        // 2. Вычисляем стартовые позиции
        int[] positions = new int[256];
        int total = 0;
        for (int i = 0; i < 256; i++) {
            positions[i] = total;
            total += counts[i];
        }
        
        // 3. Строим массив next (LF-маппинг)
        int[] next = new int[n];
        int[] tempCounts = new int[256];
        
        for (int i = 0; i < n; i++) {
            int byteIdx = bwt[i] & 0xFF;
            int pos = positions[byteIdx] + tempCounts[byteIdx];
            next[pos] = i;
            tempCounts[byteIdx]++;
        }
        
        // 4. Восстанавливаем строку
        byte[] result = new byte[n];
        int idx = bwtResult.originalIdx;
        
        for (int i = 0; i < n; i++) {
            idx = next[idx];
            result[i] = bwt[idx];
        }
        
        return result;
    }
    
    // ==================== Huffman ====================
    private static class HuffmanNode implements Comparable<HuffmanNode> {
        int frequency;
        Byte byteVal;
        HuffmanNode left;
        HuffmanNode right;
        
        HuffmanNode(int frequency, Byte byteVal) {
            this.frequency = frequency;
            this.byteVal = byteVal;
        }
        
        @Override
        public int compareTo(HuffmanNode other) {
            return Integer.compare(this.frequency, other.frequency);
        }
        
        boolean isLeaf() {
            return left == null && right == null;
        }
    }
    
    private HuffmanNode buildHuffmanTree(int[] frequencies) {
        PriorityQueue<HuffmanNode> heap = new PriorityQueue<>();
        
        for (int i = 0; i < frequencies.length; i++) {
            if (frequencies[i] > 0) {
                heap.offer(new HuffmanNode(frequencies[i], (byte) i));
            }
        }
        
        if (heap.size() == 1) {
            HuffmanNode node = heap.poll();
            HuffmanNode root = new HuffmanNode(node.frequency, null);
            root.left = node;
            root.right = new HuffmanNode(0, (byte) 0);
            return root;
        }
        
        while (heap.size() > 1) {
            HuffmanNode left = heap.poll();
            HuffmanNode right = heap.poll();
            
            HuffmanNode parent = new HuffmanNode(
                left.frequency + right.frequency,
                null
            );
            parent.left = left;
            parent.right = right;
            
            heap.offer(parent);
        }
        
        return heap.poll();
    }
    
    // Оптимизация: используем boolean[] вместо String
    private void buildHuffmanCodes(HuffmanNode node, boolean[] prefix, int length,
                                  Map<Byte, boolean[]> codes) {
        if (node.byteVal != null) {
            if (length > 0 || node.byteVal != 0) {
                boolean[] code = Arrays.copyOf(prefix, length);
                codes.put(node.byteVal, code);
            }
        } else {
            if (node.left != null) {
                if (length >= prefix.length) {
                    prefix = Arrays.copyOf(prefix, prefix.length * 2);
                }
                prefix[length] = false;
                buildHuffmanCodes(node.left, prefix, length + 1, codes);
            }
            if (node.right != null) {
                if (length >= prefix.length) {
                    prefix = Arrays.copyOf(prefix, prefix.length * 2);
                }
                prefix[length] = true;
                buildHuffmanCodes(node.right, prefix, length + 1, codes);
            }
        }
    }
    
    private static class EncodedResult {
        byte[] data;
        int bitCount;
        
        EncodedResult(byte[] data, int bitCount) {
            this.data = data;
            this.bitCount = bitCount;
        }
    }
    
    // Быстрое кодирование как в Rust
    private EncodedResult huffmanEncode(byte[] data, Map<Byte, boolean[]> codes) {
        // Предварительное выделение
        byte[] result = new byte[data.length * 2]; // Максимальный размер
        int currentByte = 0;
        int bitPos = 0;
        int byteIndex = 0;
        int totalBits = 0;
        
        for (byte b : data) {
            boolean[] code = codes.get(b);
            if (code == null) {
                throw new RuntimeException("Symbol " + b + " not found");
            }
            
            for (boolean bit : code) {
                if (bit) {
                    currentByte |= 1 << (7 - bitPos);
                }
                bitPos++;
                totalBits++;
                
                if (bitPos == 8) {
                    result[byteIndex++] = (byte) currentByte;
                    currentByte = 0;
                    bitPos = 0;
                }
            }
        }
        
        if (bitPos > 0) {
            result[byteIndex++] = (byte) currentByte;
        }
        
        byte[] finalResult = new byte[byteIndex];
        System.arraycopy(result, 0, finalResult, 0, byteIndex);
        return new EncodedResult(finalResult, totalBits);
    }
    
    private byte[] huffmanDecode(byte[] encoded, HuffmanNode root, int bitCount) {
        byte[] result = new byte[bitCount / 4 + 1];
        int resultIdx = 0;
        HuffmanNode currentNode = root;
        int bitsProcessed = 0;
        
        outer: for (byte byteVal : encoded) {
            for (int bitPos = 0; bitPos < 8; bitPos++) {
                if (bitsProcessed >= bitCount) {
                    break outer;
                }
                
                boolean bit = ((byteVal >> (7 - bitPos)) & 1) == 1;
                bitsProcessed++;
                
                currentNode = bit ? currentNode.right : currentNode.left;
                
                if (currentNode.isLeaf()) {
                    if (currentNode.byteVal != 0) {
                        if (resultIdx >= result.length) {
                            byte[] newArray = new byte[result.length * 2];
                            System.arraycopy(result, 0, newArray, 0, result.length);
                            result = newArray;
                        }
                        result[resultIdx++] = currentNode.byteVal;
                    }
                    currentNode = root;
                }
            }
        }
        
        byte[] finalResult = new byte[resultIdx];
        System.arraycopy(result, 0, finalResult, 0, resultIdx);
        return finalResult;
    }
    
    // ==================== Компрессор ====================
    private static class CompressedData {
        BWTResult bwtResult;
        int[] frequencies;
        byte[] encodedBits;
        int originalBitCount;
        
        CompressedData(BWTResult bwtResult, int[] frequencies, 
                      byte[] encodedBits, int originalBitCount) {
            this.bwtResult = bwtResult;
            this.frequencies = frequencies;
            this.encodedBits = encodedBits;
            this.originalBitCount = originalBitCount;
        }
    }
    
    private CompressedData compress(byte[] data) {
        // 1. BWT преобразование
        BWTResult bwtResult = bwtTransform(data);
        
        // 2. Подсчёт частот
        int[] frequencies = new int[256];
        for (byte b : bwtResult.transformed) {
            frequencies[b & 0xFF]++;
        }
        
        // 3. Построение дерева
        HuffmanNode huffmanTree = buildHuffmanTree(frequencies);
        
        // 4. Построение кодов (используем boolean[])
        Map<Byte, boolean[]> huffmanCodes = new HashMap<>();
        buildHuffmanCodes(huffmanTree, new boolean[32], 0, huffmanCodes);
        
        // 5. Кодирование
        EncodedResult encoded = huffmanEncode(bwtResult.transformed, huffmanCodes);
        
        return new CompressedData(bwtResult, frequencies, 
                                 encoded.data, encoded.bitCount);
    }
    
    private byte[] decompress(CompressedData compressed) {
        // 1. Восстанавливаем дерево
        HuffmanNode huffmanTree = buildHuffmanTree(compressed.frequencies);
        
        // 2. Декодирование
        byte[] decoded = huffmanDecode(
            compressed.encodedBits,
            huffmanTree,
            compressed.originalBitCount
        );
        
        // 3. Обратное BWT
        BWTResult bwtResult = new BWTResult(decoded, compressed.bwtResult.originalIdx);
        
        return bwtInverse(bwtResult);
    }
    
    // ==================== Benchmark ====================
    public Compression() {
        this.iterations = getIterations();
        this.result = 0;
    }
    
    private byte[] generateTestData(int size) {
        byte[] pattern = "ABRACADABRA".getBytes();
        byte[] data = new byte[size];
        
        for (int i = 0; i < size; i++) {
            data[i] = pattern[i % pattern.length];
        }
        
        return data;
    }
    
    @Override
    public void prepare() {
        testData = generateTestData(iterations);
    }
    
    @Override
    public void run() {
        long totalChecksum = 0;
        
        for (int i = 0; i < 5; i++) {
            CompressedData compressed = compress(testData);
            byte[] decompressed = decompress(compressed);
            long checksum = Helper.checksum(decompressed);
            
            totalChecksum = (totalChecksum + compressed.encodedBits.length) & 0xFFFFFFFFL;
            totalChecksum = (totalChecksum + (int)checksum) & 0xFFFFFFFFL;
        }
        
        result = totalChecksum;
    }
    
    @Override
    public long getResult() {
        return result;
    }
}

File: rust/src/benchmarks/compression.rs
use std::collections::{BinaryHeap, HashMap};
use std::cmp::Ordering;
use super::super::{Benchmark, helper};
use crate::config_i64;

// ==================== BWT ====================
#[derive(Debug, Clone)]
struct BWTResult {
    transformed: Vec<u8>,
    original_idx: usize,
}

fn bwt_transform(input: &[u8]) -> BWTResult {
    let n = input.len();
    if n == 0 {
        return BWTResult {
            transformed: Vec::new(),
            original_idx: 0,
        };
    }

    // 1. Создаём суффиксный массив
    let mut sa: Vec<usize> = (0..n).collect();

    // 2. Фаза 0: сортировка по первому символу (Radix sort)
    let mut buckets: Vec<Vec<usize>> = vec![Vec::new(); 256];
    
    for &idx in &sa {
        let first_char = input[idx];
        buckets[first_char as usize].push(idx);
    }

    let mut pos = 0;
    for bucket in &buckets {
        for &idx in bucket {
            sa[pos] = idx;
            pos += 1;
        }
    }

    // 3. Фаза 1: сортировка по парам символов
    if n > 1 {
        // Присваиваем ранги по первому символу
        let mut rank = vec![0; n];
        let mut current_rank = 0;
        let mut prev_char = input[sa[0]];
        
        for i in 0..n {
            let idx = sa[i];
            if input[idx] != prev_char {
                current_rank += 1;
                prev_char = input[idx];
            }
            rank[idx] = current_rank;
        }

        // Сортируем по парам (ранг[i], ранг[i+1])
        let mut k = 1;
        while k < n {
            // Создаём пары
            let mut pairs = vec![(0, 0); n];
            for i in 0..n {
                pairs[i] = (rank[i], rank[(i + k) % n]);
            }

            // Сортируем индексы по парам
            sa.sort_by(|&a, &b| {
                let pa = pairs[a];
                let pb = pairs[b];
                if pa.0 != pb.0 {
                    pa.0.cmp(&pb.0)
                } else {
                    pa.1.cmp(&pb.1)
                }
            });

            // Обновляем ранги
            let mut new_rank = vec![0; n];
            new_rank[sa[0]] = 0;
            for i in 1..n {
                let prev_pair = pairs[sa[i - 1]];
                let curr_pair = pairs[sa[i]];
                new_rank[sa[i]] = new_rank[sa[i - 1]]
                    + if prev_pair != curr_pair { 1 } else { 0 };
            }

            rank = new_rank;
            k *= 2;
        }
    }

    // 4. Собираем BWT результат
    let mut transformed = Vec::with_capacity(n);
    let mut original_idx = 0;

    for (i, &suffix) in sa.iter().enumerate() {
        if suffix == 0 {
            transformed.push(input[n - 1]);
            original_idx = i;
        } else {
            transformed.push(input[suffix - 1]);
        }
    }

    BWTResult {
        transformed,
        original_idx,
    }
}

fn bwt_inverse(bwt_result: &BWTResult) -> Vec<u8> {
    let bwt = &bwt_result.transformed;
    let original_idx = bwt_result.original_idx;
    let n = bwt.len();
    
    if n == 0 {
        return Vec::new();
    }
    
    // 1. Подсчитываем частоты символов
    let mut counts = [0usize; 256];
    for &byte in bwt {
        counts[byte as usize] += 1;
    }
    
    // 2. Вычисляем стартовые позиции для каждого символа
    let mut positions = [0usize; 256];
    let mut total = 0;
    for i in 0..256 {
        positions[i] = total;
        total += counts[i];
    }
    
    // 3. Строим массив next (LF-маппинг)
    let mut next = vec![0usize; n];
    let mut temp_counts = [0usize; 256];
    
    for (i, &byte) in bwt.iter().enumerate() {
        let byte_idx = byte as usize;
        let pos = positions[byte_idx] + temp_counts[byte_idx];
        next[pos] = i;
        temp_counts[byte_idx] += 1;
    }
    
    // 4. Восстанавливаем исходную строку
    let mut result = Vec::with_capacity(n);
    let mut idx = original_idx;
    
    for _ in 0..n {
        idx = next[idx];
        result.push(bwt[idx]);
    }
    
    result
}

// ==================== Huffman ====================
#[derive(Debug, Eq, PartialEq)]
struct HuffmanNode {
    frequency: u32,
    byte: Option<u8>,
    left: Option<Box<HuffmanNode>>,
    right: Option<Box<HuffmanNode>>,
}

impl Ord for HuffmanNode {
    fn cmp(&self, other: &Self) -> Ordering {
        // Для min-heap: меньшая частота = выше приоритет
        other.frequency.cmp(&self.frequency)
    }
}

impl PartialOrd for HuffmanNode {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}

fn build_huffman_tree(frequencies: &[u32; 256]) -> Option<Box<HuffmanNode>> {
    let mut heap = BinaryHeap::new();
    
    // Добавляем все символы с ненулевой частотой
    for (byte, &freq) in frequencies.iter().enumerate() {
        if freq > 0 {
            heap.push(Box::new(HuffmanNode {
                frequency: freq,
                byte: Some(byte as u8),
                left: None,
                right: None,
            }));
        }
    }
    
    // Если только один символ, создаём искусственный узел
    if heap.len() == 1 {
        let node = heap.pop().unwrap();
        return Some(Box::new(HuffmanNode {
            frequency: node.frequency,
            byte: None,
            left: Some(node),
            right: Some(Box::new(HuffmanNode {
                frequency: 0,
                byte: Some(0), // фиктивный символ
                left: None,
                right: None,
            })),
        }));
    }
    
    // Строим дерево
    while heap.len() > 1 {
        let left = heap.pop().unwrap();
        let right = heap.pop().unwrap();
        
        let parent = Box::new(HuffmanNode {
            frequency: left.frequency + right.frequency,
            byte: None,
            left: Some(left),
            right: Some(right),
        });
        
        heap.push(parent);
    }
    
    heap.pop()
}

fn build_huffman_codes(node: &HuffmanNode, prefix: Vec<bool>, codes: &mut HashMap<u8, Vec<bool>>) {
    if let Some(byte) = node.byte {
        if !prefix.is_empty() || byte != 0 { // Игнорируем фиктивный символ
            codes.insert(byte, prefix);
        }
    } else {
        if let Some(ref left) = node.left {
            let mut left_prefix = prefix.clone();
            left_prefix.push(false);
            build_huffman_codes(left, left_prefix, codes);
        }
        
        if let Some(ref right) = node.right {
            let mut right_prefix = prefix;
            right_prefix.push(true);
            build_huffman_codes(right, right_prefix, codes);
        }
    }
}

fn huffman_encode(data: &[u8], codes: &HashMap<u8, Vec<bool>>) -> (Vec<u8>, usize) {
    // Сначала собираем все биты
    let mut bits = Vec::new();
    for &byte in data {
        if let Some(code) = codes.get(&byte) {
            bits.extend(code);
        } else {
            // Если символ не найден (не должен случаться)
            panic!("Symbol {:?} not found in Huffman codes", byte);
        }
    }
    
    // Упаковываем биты в байты
    let mut result = Vec::with_capacity((bits.len() + 7) / 8);
    for chunk in bits.chunks(8) {
        let mut byte = 0u8;
        for (i, &bit) in chunk.iter().enumerate() {
            if bit {
                byte |= 1 << (7 - i);
            }
        }
        result.push(byte);
    }
    
    (result, bits.len())
}

fn huffman_decode(encoded: &[u8], root: &HuffmanNode, bit_count: usize) -> Vec<u8> {
    let mut result = Vec::new();
    let mut current_node = root;
    let mut bits_processed = 0;
    
    'outer: for &byte in encoded {
        for bit_pos in 0..8 {
            if bits_processed >= bit_count {
                break 'outer;
            }
            
            // Читаем бит (старший бит first)
            let bit = (byte >> (7 - bit_pos)) & 1 == 1;
            bits_processed += 1;
            
            // Идём по дереву
            current_node = if bit {
                current_node.right.as_ref().unwrap()
            } else {
                current_node.left.as_ref().unwrap()
            };
            
            // Если достигли листа
            if let Some(byte_val) = current_node.byte {
                if byte_val != 0 { // Игнорируем фиктивный символ
                    result.push(byte_val);
                }
                current_node = root; // Возвращаемся к корню
            }
        }
    }
    
    result
}

// ==================== Компрессор ====================
#[derive(Debug, Clone)]
struct CompressedData {
    bwt_result: BWTResult,
    frequencies: [u32; 256],
    encoded_bits: Vec<u8>,
    original_bit_count: usize,
}

fn compress(data: &[u8]) -> CompressedData {
    // 1. BWT преобразование
    let bwt_result = bwt_transform(data);
    
    // 2. Подсчёт частот для Huffman
    let mut frequencies = [0u32; 256];
    for &byte in &bwt_result.transformed {
        frequencies[byte as usize] += 1;
    }
    
    // 3. Построение дерева Huffman
    let huffman_tree = build_huffman_tree(&frequencies).unwrap();
    
    // 4. Построение кодов
    let mut huffman_codes = HashMap::new();
    build_huffman_codes(&huffman_tree, Vec::new(), &mut huffman_codes);
    
    // 5. Кодирование
    let (encoded_bits, bit_count) = huffman_encode(&bwt_result.transformed, &huffman_codes);
    
    CompressedData {
        bwt_result,
        frequencies,
        encoded_bits,
        original_bit_count: bit_count,
    }
}

fn decompress(compressed: &CompressedData) -> Vec<u8> {
    // 1. Восстанавливаем дерево Huffman из сохранённых частот
    let huffman_tree = build_huffman_tree(&compressed.frequencies).unwrap();
    
    // 2. Декодирование Huffman
    let decoded = huffman_decode(
        &compressed.encoded_bits,
        &huffman_tree,
        compressed.original_bit_count
    );
    
    // 3. Обратное BWT
    let bwt_result = BWTResult {
        transformed: decoded,
        original_idx: compressed.bwt_result.original_idx,
    };
    
    bwt_inverse(&bwt_result)
}

// ==================== Бенчмарк Compression ====================
pub struct Compression {
    size_val: i64,
    test_data: Vec<u8>,
    result_val: u32,
}

impl Compression {
    pub fn new() -> Self {
        let size_val = config_i64("Compression", "size");
        
        Self {
            size_val,
            test_data: Vec::new(),
            result_val: 0,
        }
    }
    
    fn generate_test_data(&self, size: i64) -> Vec<u8> {
        // Простой паттерн для хорошего сжатия
        let pattern = b"ABRACADABRA";
        let mut data = Vec::with_capacity(size as usize);
        
        for i in 0..size {
            data.push(pattern[(i as usize) % pattern.len()]);
        }
        
        data
    }
}

impl Benchmark for Compression {
    fn name(&self) -> String {
        "Compression".to_string()
    }
    
    fn prepare(&mut self) {
        self.test_data = self.generate_test_data(self.size_val);
    }

    fn run(&mut self, _iteration_id: i64) {
        let compressed = compress(&self.test_data);
        self.result_val = self.result_val.wrapping_add(compressed.encoded_bits.len() as u32);
    }
    
    fn checksum(&self) -> u32 {
        self.result_val
    }
}

// ==================== Бенчмарк Decompression ====================
pub struct Decompression {
    compression_base: Compression,
    compressed_data: Option<CompressedData>,
    decompressed: Vec<u8>,
}

impl Decompression {
    pub fn new() -> Self {
        // Создаем Compression с правильным size_val
        let size_val = config_i64("Decompression", "size");  // Используем "Decompression", "size"
        
        let mut compression_base = Compression {
            size_val,
            test_data: Vec::new(),
            result_val: 0,
        };
        
        // Генерируем тестовые данные
        compression_base.test_data = compression_base.generate_test_data(size_val);
        
        Self {
            compression_base,
            compressed_data: None,
            decompressed: Vec::new(),
        }
    }
}

impl Benchmark for Decompression {
    fn name(&self) -> String {
        "Decompression".to_string()
    }
    
    fn prepare(&mut self) {
        // НЕ вызываем compression_base.prepare() - мы уже создали test_data в конструкторе
        // Сжимаем тестовые данные
        let test_data = &self.compression_base.test_data;
        self.compressed_data = Some(compress(test_data));
        
        // Сбрасываем result_val на всякий случай
        self.compression_base.result_val = 0;
    }
    
    fn run(&mut self, _iteration_id: i64) {
        // Распаковываем данные
        if let Some(ref compressed) = self.compressed_data {
            self.decompressed = decompress(compressed);
            // Добавляем размер распакованных данных к результату
            self.compression_base.result_val = self.compression_base.result_val
                .wrapping_add(self.decompressed.len() as u32);
        }
    }
    
    fn checksum(&self) -> u32 {
        let mut res = self.compression_base.result_val;
        
        // Проверяем, что распакованные данные совпадают с исходными
        if self.compressed_data.is_some() && 
           self.decompressed == self.compression_base.test_data {
            res = res.wrapping_add(1000000);
        }
        
        res
    }
}

unsafe impl Send for Compression {}
unsafe impl Sync for Compression {}
unsafe impl Send for Decompression {}
unsafe impl Sync for Decompression {}